File: hospital/AMarotoChivite-HospitalTest/AMarotoChivite-HospitalTest-Java/src/main/java/HospitalCollections/controllers/PatientController.java
package HospitalCollections.controllers;

import HospitalCollections.base.IHospitalRepository;
import HospitalCollections.exceptions.FullRepositoryException;
import HospitalCollections.exceptions.PatientBadRequestException;
import HospitalCollections.exceptions.PatientNotFoundException;
import HospitalCollections.models.Patient;
import HospitalCollections.repositories.PatientListRepository;
import HospitalCollections.repositories.PatientMapRepository;
import HospitalCollections.repositories.PatientSetRepository;

import java.util.List;

public class PatientController implements IHospitalRepository {
    private final PatientSetRepository patientSetRepository;
    private final PatientMapRepository patientMapRepository;
    private final PatientListRepository patientListRepository;
    private TypeRepo typeRepo;

    public PatientController(PatientListRepository patientListRepository, PatientSetRepository patientSetRepository, PatientMapRepository patientMapRepository) {
        this.patientListRepository = patientListRepository;
        this.patientSetRepository = patientSetRepository;
        this.patientMapRepository = patientMapRepository;
    }

    private static Boolean filterDni(String dni) {
        String regex = ("[0-9]{8}[A-Za-z]");
        return dni.matches(regex);
    }

    private static boolean filterName(String name) {
        String regex = "[a-zA-Z ]{3,20}";
        return name.matches(regex);
    }

    private static boolean filterDateIso(String dateIso) {
        String regex = "\\d{4}-\\d{2}-\\d{2}";
        return dateIso.matches(regex);
    }

    private void checkMaxCapacity() throws FullRepositoryException {
        switch (this.typeRepo) {
            case LIST -> {
                if (patientListRepository.isFull()) {
                    throw new FullRepositoryException("La base de datos está llena, no se pueden guardar más pacientes.");
                }
            }
            case SET -> {
                if (patientSetRepository.isFull()) {
                    throw new FullRepositoryException("La base de datos está llena, no se pueden guardar más pacientes.");
                }
            }
            case MAP -> {
                if (patientMapRepository.isFull()) {
                    throw new FullRepositoryException("La base de datos está llena, no se pueden guardar más pacientes.");
                }
            }
        }
    }

    /*
    Para cambiar el tipo de repositorio que controlaremos
     */
    public void setTypeRepo(TypeRepo typeRepo) {
        this.typeRepo = typeRepo;
    }

    public TypeRepo getTypeRepo(TypeRepo typeRepo) {
        return this.typeRepo;
    }

    @Override
    public List<Patient> getAll() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.getAll();
            case SET -> patientSetRepository.getAll();
            case MAP -> patientMapRepository.getAll();
        };
    }

    @Override
    public Patient getById(String dni) throws PatientNotFoundException {
        Patient patient = null;
        switch (this.typeRepo) {
            case LIST -> patient = patientListRepository.getById(dni);
            case SET -> patient = patientSetRepository.getById(dni);
            case MAP -> patient = patientMapRepository.getById(dni);
        }
        if (patient == null) {
            throw new PatientNotFoundException("El paciente con DNI " + dni + " no se encuentra en la base de datos.");
        }
        return patient;
    }

    @Override
    public Patient save(Patient entity) throws PatientBadRequestException, FullRepositoryException {
        // Comprobamos si tiene los datos correctors
        checkCorrectPatient(entity);

        // Comprobamos que se puedan introducir más objetos
        checkMaxCapacity();

        return switch (this.typeRepo) {
            case LIST -> patientListRepository.save(entity);
            case SET -> patientSetRepository.save(entity);
            case MAP -> patientMapRepository.save(entity);
        };
    }

    @Override
    public Patient deleteById(String dni) throws PatientNotFoundException {
        Patient patient = null;
        switch (this.typeRepo) {
            case LIST -> patient = patientListRepository.deleteById(dni);
            case SET -> patient = patientSetRepository.deleteById(dni);
            case MAP -> patient = patientMapRepository.deleteById(dni);
        }
        if (patient == null) {
            throw new PatientNotFoundException("El paciente con DNI " + dni + " no se encuentra en la base de datos.");
        }
        return patient;
    }

    @Override
    public boolean isFull() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.isFull();
            case SET -> patientSetRepository.isFull();
            case MAP -> patientMapRepository.isFull();
        };
    }

    @Override
    public int getNumPacients() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.getNumPacients();
            case SET -> patientSetRepository.getNumPacients();
            case MAP -> patientMapRepository.getNumPacients();
        };
    }

    @Override
    public int[] getNumByType() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.getNumByType();
            case SET -> patientSetRepository.getNumByType();
            case MAP -> patientMapRepository.getNumByType();
        };
    }

    @Override
    public List<Patient> getAllOrderByEnterDate() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.getAllOrderByEnterDate();
            case SET -> patientSetRepository.getAllOrderByEnterDate();
            case MAP -> patientMapRepository.getAllOrderByEnterDate();
        };
    }

    @Override
    public List<Patient> getAllOrderByName() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.getAllOrderByName();
            case SET -> patientSetRepository.getAllOrderByName();
            case MAP -> patientMapRepository.getAllOrderByName();
        };
    }

    @Override
    public List<Patient> getAllOrderByType() {
        return switch (this.typeRepo) {
            case LIST -> patientListRepository.getAllOrderByType();
            case SET -> patientSetRepository.getAllOrderByType();
            case MAP -> patientMapRepository.getAllOrderByType();
        };
    }

    private void checkCorrectPatient(Patient patient) throws PatientBadRequestException {
        if (patient.getDni() == null || patient.getDni().equals("")) {
            throw new PatientBadRequestException("El DNI no puede estar vacío");
        }
        if (!filterDni(patient.getDni())) {
            throw new PatientBadRequestException("El DNI debe tener 8 número y una letra: (12345678A)!");
        }
        if (patient.getName() == null || patient.getName().equals("")) {
            throw new PatientBadRequestException("El nombre no puede estar vacío");
        }
        if (!filterName(patient.getName())) {
            throw new PatientBadRequestException("El nombre debe tener entre 3 y 20 caracteres");
        }
        if (patient.getDateNac() == null || patient.getDateNac().equals("")) {
            throw new PatientBadRequestException("La fecha de nacimiento no puede estar vacía");
        }
        if (!filterDateIso(patient.getDateNac())) {
            throw new PatientBadRequestException("La fecha debe estar en formato (aaaa-mm-dd)");
        }
        if (patient.getType() == null) {
            throw new PatientBadRequestException("El tipo del paciente no puede estar vacío");
        }
        if (patient.getDateEnter() == null || patient.getDateEnter().equals("")) {
            throw new PatientBadRequestException("La fecha de entrada no puede estar vacía");
        }
        if (!filterDateIso(patient.getDateEnter())) {
            throw new PatientBadRequestException("La fecha debe estar en formato (aaaa-mm-dd)");
        }
    }

    public enum TypeRepo {
        LIST,
        SET,
        MAP
    }
}

File: hospital/AMarotoChivite-HospitalTest/AMarotoChivite-HospitalTest-Kotlin/src/main/kotlin/HospitalCollections/controllers/PatientController.kt
package HospitalCollections.controllers

import HospitalCollections.base.IHospitalRepository
import HospitalCollections.exceptions.FullRepositoryException
import HospitalCollections.exceptions.PatientBadRequestException
import HospitalCollections.exceptions.PatientNotFoundException
import HospitalCollections.exceptions.TypeRepositoryException
import HospitalCollections.models.Patient
import HospitalCollections.repositories.PatientListRepository
import HospitalCollections.repositories.PatientMapRepository
import HospitalCollections.repositories.PatientSetRepository

class PatientController(
    private val patientListRepository: PatientListRepository?,
    private val patientSetRepository: PatientSetRepository?,
    private val patientMapRepository: PatientMapRepository?
) : IHospitalRepository {

    private var typeRepo: TypeRepo? = null

    private fun checkMaxCapacity(typeRepo: TypeRepo?) {

        when (typeRepo) {
            TypeRepo.LIST -> {
                if (patientListRepository!!.isFull()) {
                    throw FullRepositoryException("La base de datos está llena, no se pueden guardar más pacientes.")
                }
            }

            TypeRepo.SET -> {
                if (patientSetRepository!!.isFull()) {
                    throw FullRepositoryException("La base de datos está llena, no se pueden guardar más pacientes.")
                }
            }

            TypeRepo.MAP -> {
                if (patientMapRepository!!.isFull()) {
                    throw FullRepositoryException("La base de datos está llena, no se pueden guardar más pacientes.")
                }
            }

            null -> throw FullRepositoryException("La base no existe!")

        }
    }

    /*
    Para cambiar el tipo de repositorio que controlaremos
     */
    fun setTypeRepo(typeRepo: TypeRepo?) {
        this.typeRepo = typeRepo
    }

    fun getTypeRepo(typeRepo: TypeRepo?): TypeRepo? {
        return this.typeRepo
    }

    override fun getAll(): List<Patient?> {
        return when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.getAll()
            TypeRepo.SET -> patientSetRepository!!.getAll()
            TypeRepo.MAP -> patientMapRepository!!.getAll()
            null -> throw PatientNotFoundException("No existe ningún paciente")
        }
    }

    override fun getById(id: String?): Patient {
        var patient: Patient? = null
        when (typeRepo) {
            TypeRepo.LIST -> {
                patient = patientListRepository!!.getById(id)
            }

            TypeRepo.SET -> {
                patient = patientSetRepository!!.getById(id)

            }

            TypeRepo.MAP -> {
                patient = patientMapRepository!!.getById(id)
            }

            else -> {
                throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
            }
        }
        if (patient == null) {
            throw PatientNotFoundException("El paciente con DNI $id no se encuentra en la base de datos.")
        }
        return patient
    }

    override fun save(entity: Patient?): Patient? {
        // Comprobamos si tiene los datos correctors
        if (entity == null) {
            throw PatientNotFoundException("No has asignado ningún paciente para guardar")
        }
        checkCorrectPatient(entity)

        // Comprobamos que se puedan introducir más objetos
        checkMaxCapacity(typeRepo)
        return when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.save(entity)
            TypeRepo.SET -> patientSetRepository!!.save(entity)
            TypeRepo.MAP -> patientMapRepository!!.save(entity)
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
    }

    override fun deleteById(id: String?): Patient {
        var patient: Patient? = null
        when (typeRepo) {
            TypeRepo.LIST -> patient = patientListRepository!!.deleteById(id)
            TypeRepo.SET -> patient = patientSetRepository!!.deleteById(id)
            TypeRepo.MAP -> patient = patientMapRepository!!.deleteById(id)
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
        if (patient == null) {
            throw PatientNotFoundException("El paciente con DNI $id no se encuentra en la base de datos.")
        }
        return patient
    }

    override fun isFull(): Boolean {
        return when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.isFull()
            TypeRepo.SET -> patientSetRepository!!.isFull()
            TypeRepo.MAP -> patientMapRepository!!.isFull()
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
    }

    override val numPacients: Int
        get() = when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.numPacients
            TypeRepo.SET -> patientSetRepository!!.numPacients
            TypeRepo.MAP -> patientMapRepository!!.numPacients
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
    override val numByType: IntArray
        get() = when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.numByType
            TypeRepo.SET -> patientSetRepository!!.numByType
            TypeRepo.MAP -> patientMapRepository!!.numByType
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
    override val allOrderByEnterDate: List<Patient?>
        get() = when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.allOrderByEnterDate
            TypeRepo.SET -> patientSetRepository!!.allOrderByEnterDate
            TypeRepo.MAP -> patientMapRepository!!.allOrderByEnterDate
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
    override val allOrderByName: List<Patient?>
        get() = when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.allOrderByName
            TypeRepo.SET -> patientSetRepository!!.allOrderByName
            TypeRepo.MAP -> patientMapRepository!!.allOrderByName
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }
    override val allOrderByType: List<Patient?>
        get() = when (typeRepo) {
            TypeRepo.LIST -> patientListRepository!!.allOrderByType
            TypeRepo.SET -> patientSetRepository!!.allOrderByType
            TypeRepo.MAP -> patientMapRepository!!.allOrderByType
            null -> throw TypeRepositoryException("Debes asignar un tipo de repositorio!")
        }

    private fun checkCorrectPatient(patient: Patient) {
        if (patient.getDni() == null || patient.getDni() == "") {
            throw PatientBadRequestException("El DNI no puede estar vacío")
        }
        if (!filterDni(patient.getDni())) {
            throw PatientBadRequestException("El DNI debe tener 8 número y una letra: (12345678A)!")
        }
        if (patient.getName() == null || patient.getName() == "") {
            throw PatientBadRequestException("El nombre no puede estar vacío")
        }
        if (!filterName(patient.getName())) {
            throw PatientBadRequestException("El nombre debe tener entre 3 y 20 caracteres")
        }
        if (patient.getDateNac() == null || patient.getDateNac() == "") {
            throw PatientBadRequestException("La fecha de nacimiento no puede estar vacía")
        }
        if (!filterDateIso(patient.getDateNac())) {
            throw PatientBadRequestException("La fecha debe estar en formato (aaaa-mm-dd)")
        }
        if (patient.type == null) {
            throw PatientBadRequestException("El tipo del paciente no puede estar vacío")
        }
        if (patient.getDateEnter() == null || patient.getDateEnter() == "") {
            throw PatientBadRequestException("La fecha de entrada no puede estar vacía")
        }
        if (!filterDateIso(patient.getDateEnter())) {
            throw PatientBadRequestException("La fecha debe estar en formato (aaaa-mm-dd)")
        }
    }

    enum class TypeRepo {
        LIST, SET, MAP
    }

    companion object {
        private fun filterDni(dni: String?): Boolean {
            val regex = "[0-9]{8}[A-Za-z]"
            return dni!!.matches(regex.toRegex())
        }

        private fun filterName(name: String?): Boolean {
            val regex = "[a-zA-Z ]{3,20}"
            return name!!.matches(regex.toRegex())
        }

        private fun filterDateIso(dateIso: String?): Boolean {
            val regex = "\\d{4}-\\d{2}-\\d{2}"
            return dateIso!!.matches(regex.toRegex())
        }
    }
}